### PR TITLE
[RISCV][NFC] Simplify RISCVInstrInfo::copyPhysReg

### PR DESCRIPTION
**Pull request by wangpc-pp**
Add a macro to reduce duplicated codes of RVV copies.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 1]
```diff
@@ -473,122 +473,48 @@ void RISCVInstrInfo::copyPhysReg(MachineBasicBlock &MBB,
...
-      RISCV::GPRRegClass.contains(SrcReg)) {
-    assert(STI.getXLen() == 64 && "Unexpected GPR size");
-    BuildMI(MBB, MBBI, DL, get(RISCV::FMV_D_X), DstReg)
-        .addReg(SrcReg, getKillRegState(KillSrc));
-    return;
-  }
-
-  if (RISCV::GPRRegClass.contains(DstReg) &&
-      RISCV::FPR64RegClass.contains(SrcReg)) {
-    assert(STI.getXLen() == 64 && "Unexpected GPR size");
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:500]: This loses the asserts?

I don't think the scalar registers have that much duplication. What would you think about just deduplicating the vector registers instead?

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 2]
```diff
@@ -473,122 +473,48 @@ void RISCVInstrInfo::copyPhysReg(MachineBasicBlock &MBB,
...
+  COPY_FROM_TO(RISCV::FPR64RegClass, RISCV::GPRRegClass, RISCV::FMV_X_D);
+
+#define RVV_COPY_IMPL(REG_CLASS, LMUL, NF)                                     \
+  do {                                                                         \
+    if (RISCV::REG_CLASS.contains(DstReg, SrcReg)) {                           \
+      copyPhysRegVector(MBB, MBBI, DL, DstReg, SrcReg, KillSrc,                \
+                        RISCV::VMV##LMUL##R_V, NF);                            \
+      return;                                                                  \
+    }                                                                          \
+  } while (0)
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:497]: Alternatively we could just use a lambda. I think I would personally prefer to see the full `RISCV::VM2R_V` register/`VRN2M4RegClass` regclass spelt out.

Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:497]: How can we return from a lambda?

Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:497]: Woops you're right, I completely glossed over the early exit.

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 3]
```diff
@@ -473,122 +473,48 @@ void RISCVInstrInfo::copyPhysReg(MachineBasicBlock &MBB,
...
+      BuildMI(MBB, MBBI, DL, get(OPC), DstReg)                                 \
+          .addReg(SrcReg, getKillRegState(KillSrc));                           \
+      return;                                                                  \
+    }                                                                          \
+  } while (0)
+
+  COPY_FROM_TO(RISCV::GPRRegClass, RISCV::FPR32RegClass, RISCV::FMV_W_X);
+  COPY_FROM_TO(RISCV::FPR32RegClass, RISCV::GPRRegClass, RISCV::FMV_X_W);
+  COPY_FROM_TO(RISCV::GPRRegClass, RISCV::FPR64RegClass, RISCV::FMV_D_X);
+  COPY_FROM_TO(RISCV::FPR64RegClass, RISCV::GPRRegClass, RISCV::FMV_X_D);
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:488]: Please undef the macros when you're done with them.

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp (hunk 2 above)
Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp]: I think macros is more intuitive. :-)

**Review by lukel97 — APPROVED** [review]

**Comment by jrtc27** [comment]
Why is this not just a loop over an array of triples? Macros should be avoided in C++ where possible, and they’re not needed here.

**Comment by wangpc-pp** [comment]
> Why is this not just a loop over an array of triples? Macros should be avoided in C++ where possible, and they’re not needed here.

So what need I revise them to use a loop? I think we have a lot of macros even in target-independent code.

**Comment by topperc** [comment]
Is your goal just to reduce the duplicated code or do you have other changes planned in this area?

I'd like to see us use larger LMUL moves when possible in the tuple copies. VRN8M1 copy doesn't need to use 8 instructions if the registers are LMUL8 aligned.

**Comment by wangpc-pp** [comment]
Closed as there is a better alternative.

**Pull request closed without merging**